pkg/chunk/aws/metrics_autoscaling.go

```diff
@@ -0,0 +1,218 @@
+	if newWrite > expected.WriteScale.MaxCapacity {
+		newWrite = expected.WriteScale.MaxCapacity
+	}
+	earliest := d.metrics.tableLastUpdated[current.Name].Add(time.Duration(expected.WriteScale.OutCooldown) * time.Second)
+	if earliest.After(mtime.Now()) {
+		level.Info(util.Logger).Log("msg", "deferring "+msg, "table", current.Name, "till", earliest)
+		return
+	}
+	if newWrite > current.ProvisionedWrite {
+		level.Info(util.Logger).Log("msg", msg, "table", current.Name, "write", newWrite)
+		expected.ProvisionedWrite = newWrite
+		d.metrics.tableLastUpdated[current.Name] = mtime.Now()
+	}
+}
+
+func newMetrics(cfg DynamoDBConfig) (*metricsData, error) {
+	var promAPI promV1.API
+	if cfg.MetricsURL != "" {
+		client, err := promApi.NewClient(promApi.Config{Address: cfg.MetricsURL})
+		if err != nil {
+			return nil, err
+		}
+		promAPI = promV1.NewAPI(client)
+	}
+	return &metricsData{
+		promAPI:           promAPI,
+		queueLengthTarget: cfg.MetricsTargetQueueLen,
+		scaleUpFactor:     cfg.MetricsScaleUpFactor,
+		tableLastUpdated:  make(map[string]time.Time),
+	}, nil
+}
+
+func (m *metricsData) update(ctx context.Context) error {
+	if m.promLastQuery.After(mtime.Now().Add(-cachePromDataFor)) {
+		return nil
+	}
+
+	m.promLastQuery = mtime.Now()
+	// average the queue length over 2 minutes to avoid aliasing with the 1-minute flush period
+	// TODO: adjust that 2m depending on configuration of the flush period
+	qlMatrix, err := promQuery(ctx, m.promAPI, `sum(avg_over_time(cortex_ingester_flush_queue_length{job="cortex/ingester"}[2m]))`, queueObservationPeriod, queueObservationPeriod/2)
+	if err != nil {
+		return err
+	}
+	if len(qlMatrix) != 1 {
+		return errors.Errorf("expected one sample stream for queue: %d", len(qlMatrix))
+	}
+	if len(qlMatrix[0].Values) != 3 {
+		return errors.Errorf("expected three values: %d", len(qlMatrix[0].Values))
+	}
+	m.queueLengths = make([]float64, len(qlMatrix[0].Values))
+	for i, v := range qlMatrix[0].Values {
+		m.queueLengths[i] = float64(v.Value)
+	}
+
+	// fetch write error rate per DynamoDB table
+	deMatrix, err := promQuery(ctx, m.promAPI, `sum(rate(cortex_dynamo_failures_total{error="ProvisionedThroughputExceededException",operation=~".*Write.*"}[1m])) by (table) > 0`, 0, time.Second)
+	if err != nil {
+		return err
+	}
+	if m.errorRates, err = extractRates(deMatrix); err != nil {
+		return err
+	}
+
+	// fetch write capacity usage per DynamoDB table
+	// use the rate over 15 minutes so we take a broad average
+	usageMatrix, err := promQuery(ctx, m.promAPI, `sum(rate(cortex_dynamo_consumed_capacity_total{operation="DynamoDB.BatchWriteItem"}[15m])) by (table) > 0`, 0, time.Second)
+	if err != nil {
+		return err
+	}
+	if m.usageRates, err = extractRates(usageMatrix); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func extractRates(matrix model.Matrix) (map[string]float64, error) {
+	ret := make(map[string]float64)
```

Comment:
IIRC throughout the codebase we tend to prefer `map[string]float64`, and only use `make` when we know the size.